### PR TITLE
[FIX] website: Cannot create account

### PR DESCRIPTION
**Pull request by pyduf**
When creating an account from the website (through the link "Don't have an account?") while having been previously identified, the creation lasts a very long time and ends-up with a http 502 error (actually a timeout).

Technical note: The problem was due to a database deadlock: a record was deleted in one transaction while being written in another (_merge_visitor deletes it and _update_visitor_last_visit writes it but with in a different transaction).

Task-2950241

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/models/website_visitor.py [hunk 1]
```diff
@@ -327,6 +327,8 @@ def _merge_visitor(self, target):
 
         :param target: main visitor, target of link process;
         """
+        if self == target:
```
Comment (std-odoo) [inline addons/website/models/website_visitor.py:330]: it means that all override of `_merge_visitor` will be called
shouldn't we do this verification before calling the method ?

**Review by std-odoo — COMMENTED** [review]
@pyduf Validation :)

File: addons/website/models/website_visitor.py (hunk 1 above)
Comment (std-odoo) [reply addons/website/models/website_visitor.py]: ok for me :)

File: addons/website/models/res_users.py [hunk 2]
```diff
@@ -72,18 +72,17 @@ def authenticate(cls, db, login, password, user_agent_env):
...
-                ], limit=1)
-                if visitor_current_user_sudo:
-                    # A visitor exists for the logged in user, link public
-                    # visitor records to it.
-                    visitor_pre_authenticate_sudo._merge_visitor(visitor_current_user_sudo)
-                    visitor_current_user_sudo._update_visitor_last_visit()
-                else:
-                    visitor_pre_authenticate_sudo.access_token = user_partner.id
-                    visitor_pre_authenticate_sudo._update_visitor_last_visit()
+            env_uid = api.Environment(visitor_pre_authenticate_sudo.env.cr, uid, {})
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/res_users.py:75]: I'd not modify this line unless there is a good reason provided (and indicated in the commit message)
I don't remember well why it was like that when I refactored this part but I didn't touch it back then as it was probably there for a good reason. Sadly it didn't come with a comment or something in the commit message :shrug: :shrug: 

From what I understand, that's just mimicking what is done everywhere else after a login/authenticate, new cursor is created everywhere, that's probably to be sure everything is detached from the previous user and no issue given identify/rights.

File: addons/website/models/res_users.py [hunk 3]
```diff
@@ -72,18 +72,18 @@ def authenticate(cls, db, login, password, user_agent_env):
...
-                    # visitor records to it.
+            env_uid = api.Environment(visitor_pre_authenticate_sudo.env.cr, uid, {})
+            user_partner = env_uid.user.partner_id
+            visitor_current_user_sudo = env_uid['website.visitor'].sudo().search([
+                ('partner_id', '=', user_partner.id)
+            ], limit=1)
+            if visitor_current_user_sudo:
+                # A visitor exists for the logged in user, link public
+                # visitor records to it.
+                if visitor_pre_authenticate_sudo != visitor_current_user_sudo:
```
Comment (rdeodoo) [inline addons/website/models/res_users.py:83]: I'd just keep this line in the PR if this is enough.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Following our discussion.
Quickly tested and I couldn't reproduce it locally while I can on runbot.
Reading the code and trying to reproduce it locally, here is what I found:
- The `authenticate` method is called twice during the signup. I guess the first time the `visitor_current_user_sudo` is a brand new visitor linked to the newly created user. Always != than `visitor_pre_authenticate_sudo`. Then on the second call to `authenticate`, obviously both will be the same, as `authenticate` is being called for the user while this exact user is already logged in.
- Following first point, I thought it should also break if you try to login again when you are already login, and it seems to be the case. While I couldn't reproduce it locally, I could on runbot: Login as admin > go to /web/login > login again > 502 error.

File: addons/website/models/res_users.py (hunk 3 above)
Comment (pyduf) [reply addons/website/models/res_users.py:83]: Thank you for your help. I will only keep this line so and test again to ensure it solves the problem.

Comment (pyduf) [reply addons/website/models/res_users.py]: Just to notify that the fix still works with the minimal change (just the added "if" statement to ensure the 2 visitors to be merged are different)

Comment (rdeodoo) [reply addons/website/models/res_users.py]: Nice, just don't forget to reflect what I said in the commit message:
- This is not only about creating account but also login when already login
- This is due to going twice in the authenticate, when the second time we are already login

Kinda helpful ^^

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Pull request closed without merging**